### PR TITLE
K8s: improve frontend resource client

### PR DESCRIPTION
**Pull request by ryantxu**
This PR improces the resource client so that:

1. On create, it sets a `generateName` with the first two leters of the logged in user (rather than always "x")
2. Adds "UI" as the origin info to any operation from the UI
3. Updates `AnnoKeyOriginKey` > AnnoKeyOriginHash` 
4. improves the return type for create -- no longer void

### CODE REVIEW

File: public/app/features/apiserver/client.ts [hunk 1]
```diff
@@ -90,3 +97,12 @@ export class ScopedResourceClient<T = object, K = string> implements ResourceCli
       .join(',');
   }
 }
+
+// add the origin annotations so we know what was set from the UI
+function setOriginAsUI(meta: Partial<ObjectMeta>) {
+  if (!meta.annotations) {
+    meta.annotations = {}
+  }
+  meta.annotations.AnnoKeyOriginName = "UI"
```
Comment (toddtreece) [inline public/app/features/apiserver/client.ts:106]: would it be useful to have the grafana UI version and commit info here for debugging purposes?

Comment (ryantxu) [reply public/app/features/apiserver/client.ts:106]: good idea!

File: public/app/features/apiserver/client.ts [hunk 2]
```diff
@@ -90,3 +97,13 @@ export class ScopedResourceClient<T = object, K = string> implements ResourceCli
       .join(',');
   }
 }
+
+// add the origin annotations so we know what was set from the UI
+function setOriginAsUI(meta: Partial<ObjectMeta>) {
+  if (!meta.annotations) {
+    meta.annotations = {};
+  }
+  meta.annotations.AnnoKeyOriginName = 'UI';
```
Comment (ryantxu) [inline public/app/features/apiserver/client.ts:106]: As discussed with @fabrizia a while ago.  This way we can better identify what came from the UI vs all the other ways things can get into the system

File: public/app/features/apiserver/client.ts [hunk 3]
```diff
@@ -90,3 +97,13 @@ export class ScopedResourceClient<T = object, K = string> implements ResourceCli
...
+  meta.annotations.AnnoKeyOriginPath = window.location.pathname;
+  meta.annotations.AnnoKeyOriginHash = config.buildInfo.versionString;
```
Comment (toddtreece) [inline public/app/features/apiserver/client.ts:108]: based on [this](https://github.com/grafana/grafana/blob/399651b9ad822dc0569fbf6efcbaebce6b03bdae/pkg/apimachinery/utils/meta.go#L41), isn't `AnnoKeyOriginHash` used for the dashboard checksum?

Comment (ryantxu) [reply public/app/features/apiserver/client.ts:108]: The meaning and usage of hash will depend on the origin.  But ya weird.

For things from our existing provisioning system, the hash is the file checksum

Comment (ryantxu) [reply public/app/features/apiserver/client.ts:108]: Alternative would be appending the version to path -- I don't have a strong sense of which approach is better

Comment (ryantxu) [reply public/app/features/apiserver/client.ts:108]: (name should stay low cardinality and eventually link to origins with that name!)

**Review by toddtreece — APPROVED** [review]

File: public/app/features/apiserver/client.ts (hunk 3 above)
Comment (toddtreece) [reply public/app/features/apiserver/client.ts:108]: > meaning and usage of hash will depend on the origin

this makes sense to me as long as we'll never need version info at the same time as other uses of the hash

File: public/app/features/apiserver/client.ts (hunk 2 above)
Comment (dprokop) [inline public/app/features/apiserver/client.ts:106]: Out of curiosity, where are those annotations sent to in the API server?

**Review by dprokop — APPROVED** [review]

File: public/app/features/apiserver/client.ts (hunk 2 above)
Comment (toddtreece) [reply public/app/features/apiserver/client.ts:106]: [annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) are a standard part of the k8s object's metadata, so they are passed on to storage automatically by the apiserver when creating or updating an object

Comment (dprokop) [reply public/app/features/apiserver/client.ts:106]: thanks @toddtreece !

**Pull request merged**